File: 2225-find-players-with-zero-or-one-losses/2225-find-players-with-zero-or-one-losses.java
import java.util.*;

class Solution {
    public List<List<Integer>> findWinners(int[][] matches) {
        HashMap<Integer,Boolean> winnerHashMap = new HashMap<>();
        HashMap<Integer,Integer> loserHashMap = new HashMap<>();

        for (int[] match : matches) {
            int winner = match[0];
            int loser = match[1];

            Integer loserCount = loserHashMap.getOrDefault(loser, 0);
            loserHashMap.put(loser, loserCount + 1);
            if (winnerHashMap.containsKey(loser)) {
                winnerHashMap.remove(loser);
            }

            if (!loserHashMap.containsKey(winner)) {
                winnerHashMap.put(winner, true);
            }
        }

        List<Integer> winners = new ArrayList<>();
        List<Integer> lostOnce = new ArrayList<>();

        for (HashMap.Entry<Integer,Boolean> entry : winnerHashMap.entrySet()) {
            winners.add(entry.getKey());
        }

        for (HashMap.Entry<Integer,Integer> entry : loserHashMap.entrySet()) {
            if (entry.getValue().equals(1)) {
                lostOnce.add(entry.getKey());
            }
        }

        Collections.sort(winners);
        Collections.sort(lostOnce);

        List<List<Integer>> answer = new ArrayList<>();
        answer.add(winners);
        answer.add(lostOnce);

        return answer;
    }
}

File: 2225-find-players-with-zero-or-one-losses/2225-find-players-with-zero-or-one-losses.go
func findWinners(matches [][]int) [][]int {
	winnerHashMap := make(map[int]bool)
	loserHashMap := make(map[int]int)
	for _, v := range matches {
		winner := v[0]
		loser := v[1]

		// update loser
		// 1. increment loser count in loser hashMap
		// 2. if loser presents in winner hashMap, remove it
		loserHashMap[loser]++
		if _, ok := winnerHashMap[loser]; ok {
			delete(winnerHashMap, loser)
		}

		// update winner
		// 1. if winner does not present in loserHashMap, append it
		if _, ok := loserHashMap[winner]; !ok {
			winnerHashMap[winner] = true
		}
	}

	winners := make([]int, 0)
	lostOnce := make([]int, 0)

	for k := range winnerHashMap {
		winners = append(winners, k)
	}

	for k, v := range loserHashMap {
		if v == 1 {
			lostOnce = append(lostOnce, k)
		}
	}

	sort.Slice(winners, func(i, j int) bool {
		return winners[i] < winners[j]
	})
	sort.Slice(lostOnce, func(i, j int) bool {
		return lostOnce[i] < lostOnce[j]
	})

	return [][]int{winners, lostOnce}
}
